[CTF Rebuild] Siege rework. Catapult launch force changes. Added outposts

bunniewormy opened the pull request:
## Status

(pick one)

- **READY**: this PR is (to the best of your knowledge) ready to be incorporated into the game.

## Description

[change] catapult changes:
		players are launched with 75% original force
		other items are launched with 110% original force
[change] cata rocks do more damage to blocks 
[added] cata rocks are shieldable
[removed] spawning removed from ballista 
[change] balllista no longer costs gold
[change] ballista price reduced to 150
[change] improved cata unpacking code 
[added] added ballista bomb bolts to tth
[added, big addition] add an outpost to the siege shop. Outpost mechanics:
		5x5 outpost which is essentially a workshop, but is unpacked from a crate bought in the siege shop for 150 coins and 50 gold
        respawning with shorter spawn immunity
        storage-size storage
        can change class on it (both teams)
        also works as a tunnel (only owner team can use the tunnel; other teams can however use tunnel to travel to it)
        capturable same as ballistas

MonkeyFeats commented:
The outpost seems like a big change, it would require extensive testing to evaluate whether it is a good change or just pointless.

bunniewormy commented:
> 
> 
> The outpost seems like a big change, it would require extensive testing to evaluate whether it is a good change or just pointless.

The point of the change is for ballistas to actually be a siege machine rather than a spawn point, and for outposts to be spawn points instead of them. It doesn't make much sense for something that is supposed to shoot bolts to be a respawn point. Moreover adding an improved respawn point with a tunnel improves builders' ability to control terrain + the storage is pretty useful for this as well.

It's not that big of a change - the outposts don't really do anything that other structures in the game don't, it's just ballistas with tunnels instead of ability to shoot.

Vam-Jam commented on Entities/Industry/CTFShops/Outpost/Outpost.as:
```diff
@@ -0,0 +1,83 @@
+#include "StandardControlsCommon.as"
+#include "TunnelCommon.as"
+
+u8 respawn_immunity_time = 30 * 1; // 1 second
```
const

bunniewormy replied:
fixed

Vam-Jam commented on Entities/Items/Crate/Crate.as:
```diff
@@ -441,7 +423,30 @@ void Unpack(CBlob@ this)
+		{
+			if (blob.getName() == "outpost")
+			{
+				//pos -= Vec2f(0, map.tilesize);
```
is this suppose to be commented out, if so remove the if empty if statement

bunniewormy replied:
fixed

Vam-Jam commented on Entities/Vehicles/Catapult/Catapult.as:
```diff
@@ -233,10 +236,18 @@ void Vehicle_onFire(CBlob@ this, VehicleInfo@ v, CBlob@ bullet, const u8 _charge
 
-		bullet.setVelocity(vel);
+
+		if(bullet.hasTag("player"))
+		bullet.setVelocity(vel * player_launch_modifier);
+
+		else
+		bullet.setVelocity(vel * other_launch_modifier);
```
fix formatting

bunniewormy replied:
fixed

Vam-Jam commented on Rules/CommonScripts/RespawnSystem.as:
```diff
@@ -33,6 +33,11 @@ shared class RespawnSystem
 			newBlob.server_SetPlayer(player);
 			player.server_setTeamNum(p_info.team);
+
+			if(p_info.customImmunityTime > 0)
```
if space (

bunniewormy replied:
fixed

CatWiz commented on Entities/Items/Crate/Crate.as:
```diff
@@ -184,8 +184,7 @@ void Land(CBlob@ this)
 {
-	return (this.getName() == blob.getName())
-	    || ((blob.getShape().isStatic() || blob.hasTag("player") || blob.hasTag("projectile")) && !blob.hasTag("parachute"));
+	return (blob.getShape().isStatic() || blob.getPlayer() !is null || blob.hasTag("projectile")) && !blob.hasTag("parachute");
```
this prevents crates from colliding with each other again

bunniewormy replied:
fixed

CatWiz commented on Entities/Items/Crate/Crate.as:
```diff
@@ -739,7 +758,7 @@ CBlob@ getPlayerInside(CBlob@ this)
 }
 
-bool DumpOutItems(CBlob@ this, float pop_out_speed = 5.0f, Vec2f init_velocity = Vec2f_zero, bool dump_special = true)
+bool DumpOutItems(CBlob@ this, float pop_out_speed = 5.0f, Vec2f init_velocity = Vec2f_zero, bool dump_player = true)
```
looks like there's a bunch of unintentional changes to mine trap code

bunniewormy replied:
fixed

CatWiz commented on Entities/Items/Crate/Crate.as:
```diff
@@ -348,27 +348,10 @@ void onCommand(CBlob@ this, u8 cmd, CBitStream @params)
 			CInventory@ inv = this.getInventory();
 			u8 itemcount = inv.getItemsCount();
-
-			// Boobytrap if crate has enemy mine
-			CBlob@ mine = null;
-			for (int i = 0; i < inv.getItemsCount(); i++)
-			{
-				CBlob@ item = inv.getItem(i);
-				if (item.getName() == "mine" && item.getTeamNum() != caller.getTeamNum())
-				{
-					CBitStream params;
-					params.write_u16(caller.getNetworkID());
-					params.write_u16(item.getNetworkID());
-					this.SendCommand(this.getCommandID("boobytrap"), params);
-					return;
-				}
-			}
```
why did you just remove this?

bunniewormy replied:
Fixed, accident

mugg91 commented:
If you remove spawning from Ballista, might as well remove it from Warboat, too.

> also works as a tunnel (only owner team can use the tunnel; other teams can however use tunnel to travel to it)

If the other team can travel there, then that means the other team can just travel in with a lit keg to destroy it.

AsuMagic commented:
> If you remove spawning from Ballista, might as well remove it from Warboat, too.

Maybe, but this would make warboats pretty useless compared to longboats, as warboats suck for ramming. Either way, they are entirely unused in CTF and TTH is a bit of a second class citizen, so I don't think it is very important to change this.

AsuMagic commented:
Note that I have reworked catapult rocks a bit, which was merged into this PR. See: https://github.com/bunniewormy/Base/pull/47

> - Fine-tuned tile and blob damage a fair bit (stone damage per rock significantly increased, wood damage per rock increased even further)
> - Hopefully made rock damage more consistent than before regardless of shooting distance/force/luck
> - Visual particle overhaul (wooden particles on hitting wood, dust particles on hitting ground, stone dust particles in flight, smaller gib particle when the rock cata is destroyed) and minor audio tweaks
> - Ricochet happens quite a lot more and deals more damage due to some physics tweaks, which imo feels good, and makes cata a bit deadlier in some circumstances: e.g. behind just in front of a tower under cata fire, or being inside of a tower as rocks fall into it
> - Catapults fire 7 rocks instead of 5 (40% increase in rocks per catapult shot)
> - Rock cost is increased from 2 to 3 stone (50% more expensive)

It was tested both offline and online with no major observed issues, including with lag emulation.

AsuMagic reviewed (changes requested):
General observations or issues I've observed aside of the other comments:
- The ballista icon still shows up on the minimap, even though they aren't as critical as before (catapults don't show). Outposts don't. This may be caused by `Spawner.as` which `Ballista.cfg` uses but not `Outpost.cfg`.
- We should test other gamemodes to see if any of the respawn/ballista changes negatively affects them. TTH, for one, probably is...

bunniewormy commented on Entities/Items/Crate/Crate.as:
```diff
@@ -185,7 +185,7 @@ void Land(CBlob@ this)
 {
 	return (this.getName() == blob.getName())
-	    || ((blob.getShape().isStatic() || blob.hasTag("player") || blob.hasTag("projectile")) && !blob.hasTag("parachute"));
+		|| ((blob.getShape().isStatic() || blob.getPlayer() !is null || blob.hasTag("projectile")) && !blob.hasTag("parachute"));
```
i dont know
reverted

bunniewormy commented on Entities/Common/Includes/Costs.as:
```diff
@@ -139,6 +139,8 @@ void InitCosts()
 	CTFCosts::ballista_gold =               ReadCost(cfg, "cost_ballista_gold"              , 50);
```
ok

bunniewormy commented on Rules/CommonScripts/RespawnSystem.as:
```diff
@@ -33,6 +33,11 @@ shared class RespawnSystem
 			newBlob.server_SetPlayer(player);
 			player.server_setTeamNum(p_info.team);
+
+			if (p_info.customImmunityTime > 0)
```
done

bunniewormy commented:
removed ballista's minimap icon
replaced ballista minimap icon with outpost icon in the spritesheet 
removed Spawner.as from Ballista.cfg script list 
added Spawner.as to Outpost.cfg script list; removed createRespawnPoint() from Outpost.as file
